components/sup/src/main.rs

```diff
@@ -202,76 +202,85 @@ fn sub_term(m: &ArgMatches) -> Result<()> {
 ////////////////////////////////////////////////////////////////////////
 
 fn mgrcfg_from_matches(m: &ArgMatches) -> Result<ManagerConfig> {
-    let mut cfg = ManagerConfig::default();
-    cfg.auto_update = m.is_present("AUTO_UPDATE");
-    cfg.update_url = bldr_url(m);
-    cfg.update_channel = channel(m);
+    let mut cfg = ManagerConfig {
+        auto_update: m.is_present("AUTO_UPDATE"),
+        update_url: bldr_url(m),
+        update_channel: channel(m),
+        http_disable: m.is_present("HTTP_DISABLE"),
+        organization: m.value_of("ORGANIZATION").map(|org| org.to_string()),
+        gossip_permanent: m.is_present("PERMANENT_PEER"),
+        ring_key: get_ring_key(m)?,
+        ..Default::default()
+    };
```

Comment:
I like what you're doing here. I'd like to see it taken even further, because I much prefer to have objects initialized all at once rather than in several steps. If you did something like:
```rust
    let gossip_listen = if let Some(addr_str) = m.value_of("LISTEN_GOSSIP") {
        GossipListenAddr::from_str(addr_str)?
    } else {
        GossipListenAddr::default()
    };

    let eventsrv_group = if let Some(events) = m.value_of("EVENTS") {
        ServiceGroup::from_str(events).ok()
    } else {
        None
    };

    let ctl_listen = if let Some(addr_str) = m.value_of("LISTEN_CTL") {
        SocketAddr::from_str(addr_str).ok()
    } else {
        None
    };

    let http_listen = if let Some(addr_str) = m.value_of("LISTEN_HTTP") {
        http_gateway::ListenAddr::from_str(addr_str)?
    } else {
        http_gateway::ListenAddr::default()
    };

    Ok(ManagerConfig {
        auto_update: m.is_present("AUTO_UPDATE"),
        custom_state_path: None,
        eventsrv_group,
        update_url: bldr_url(m),
        update_channel: channel(m),
        gossip_listen,
        ctl_listen: ctl_listen.unwrap_or(protocol::ctl::default_addr()),
        http_listen,
        http_disable: m.is_present("HTTP_DISABLE"),
        gossip_peers: get_peers(m)?,
        gossip_permanent: m.is_present("PERMANENT_PEER"),
        ring_key: get_ring_key(m)?,
        organization: m.value_of("ORGANIZATION").map(|org| org.to_string()),
        watch_peer_file: m.value_of("PEER_WATCH_FILE").map(Into::into),
    })
```

You could do away with the `Default` trait for `ManagerConfig` entirely. You'd have to [remove these tests](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/mod.rs#L2018-L2054), (or migrate them to the place you're adding), but I think that would be appropriate. In fact,  I think you can eliminate the [`ManagerConfig::custom_state_path`](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/mod.rs#L131) field entirely.

Reply:
This is actually just the first step in a slightly grander refactor that I think leads more to what you're suggesting here. My plan is to move this code out of main and then leverage default values from clap to eliminate a bunch of this code. In the end we should end up with just a `ManagerConfig {...}` block without any of the `if` bits.

I thought that that would warrant it's own PR though, so I didn't include it here.